ffi/derive/src/arg.rs

```diff
@@ -0,0 +1,494 @@
+            .map(|i| syn::LitInt::new(&i.to_string(), Span::call_site()))
+            .collect();
+        let field_idents: Vec<_> = (0..node.elems.len())
+            .into_iter()
+            .map(|idx| Ident::new(&format!("{}__field_{}", arg_name, idx), Span::call_site()))
+            .collect();
+
+        let (elem_ffi_type, elem_src_to_ffi, elem_ffi_to_src) = node
+            .elems
+            .iter()
+            .enumerate()
+            .map(|(i, elem)| (field_idents[i].clone(), elem.clone()))
+            .fold(
+                <(Vec<_>, Vec<_>, Vec<_>)>::default(),
+                |mut acc, (field_ident, elem)| {
+                    let elem_mapper = Arg::new(self.self_ty, field_ident, elem, self.is_input);
+
+                    acc.0.push(elem_mapper.ffi_type);
+                    acc.1.push(elem_mapper.src_to_ffi);
+                    acc.2.push(elem_mapper.ffi_to_src);
+
+                    acc
+                },
+            );
+
+        self.ffi_type = Some(parse_quote! {iroha_ffi::Pair<#( #elem_ffi_type ),*>});
+        self.src_to_ffi = Some(parse_quote! {
+            let #arg_name = {
+                #( let #field_idents = arg.#ids; )*
+                #( #elem_src_to_ffi )*
+                iroha_ffi::Pair(#( #field_idents ),*)
+            };
+        });
+        self.ffi_to_src = Some(parse_quote! {
+            let #arg_name = {
+                #( let #field_idents = arg.#ids; )*
+                #( #elem_ffi_to_src )*
+                (#( #field_idents ),*)
+            };
+        });
+    }
+}
+
+pub fn generic_arg_types(seg: &syn::PathSegment) -> Vec<&Type> {
+    if let AngleBracketed(arguments) = &seg.arguments {
+        return arguments
+            .args
+            .iter()
+            .filter_map(|arg| {
+                if let syn::GenericArgument::Type(ty) = &arg {
+                    Some(ty)
+                } else {
+                    None
+                }
+            })
+            .collect();
+    }
+
+    abort!(seg, "Type not found in the given path segment")
+}
+
+fn generic_arg_bindings(seg: &syn::PathSegment) -> Vec<&syn::Binding> {
+    if let AngleBracketed(arguments) = &seg.arguments {
+        let mut bindings = vec![];
+
+        for arg in &arguments.args {
+            if let syn::GenericArgument::Binding(binding) = arg {
+                bindings.push(binding);
+            }
+        }
+
+        return bindings;
+    };
+
+    abort!(seg, "Binding not found in the given path segment")
+}
+
+/// Visitor for path types which replaces all occurrences of `Self` with a fully qualified type
+pub struct SelfResolver<'ast> {
+    self_ty: &'ast syn::Path,
+}
+
+impl<'ast> SelfResolver<'ast> {
+    pub fn new(self_ty: &'ast syn::Path) -> Self {
+        Self { self_ty }
+    }
+}
+
+impl VisitMut for SelfResolver<'_> {
+    fn visit_path_mut(&mut self, node: &mut syn::Path) {
+        if node.leading_colon.is_some() {
+            // NOTE: It's irrelevant
```

Comment:
What's the purpose of this `if`? I don't get it from the comment.

Reply:
I could remove it, it's there as a reminder that visitor could parse the item at hand such as `leading_colon` but it's irrelevant in the current implementation